### PR TITLE
[SPARK-30814][SQL] ALTER TABLE ... ADD COLUMN position should be able to reference columns being added 

### PR DESCRIPTION
**Pull request by imback82**
### What changes were proposed in this pull request?

In ALTER TABLE, a column in ADD COLUMNS can depend on the position of a column that is just being added. For example, for a table with the following schema:
```
root:
  - a: string
  - b: long
```
, the following should work:
```
ALTER TABLE t ADD COLUMNS (x int AFTER a, y int AFTER x) 
```
Currently, the above statement will throw an exception saying that AFTER x cannot be resolved, because x doesn't exist yet. This PR proposes to fix this issue.

### Why are the changes needed?

To fix a bug described above.

### Does this PR introduce any user-facing change?

Yes, now
```
ALTER TABLE t ADD COLUMNS (x int AFTER a, y int AFTER x) 
```
works as expected.

### How was this patch tested?

Added new tests

### CODE REVIEW

**Comment by imback82** [comment]
cc @brkyvz @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -3023,9 +3023,27 @@ class Analyzer(
   object ResolveAlterTableChanges extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
       case a @ AlterTable(_, _, t: NamedRelation, changes) if t.resolved =>
+        // Keeps track of new columns being added. Keys are normalized parent names and
+        // values are field names that belong to their parent.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Can we be more clear here? For example, if we add a column `a.b.c.d`, which part should be the key? `a.b.c` or `a.b`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 2]
```diff
@@ -483,8 +484,16 @@ trait CheckAnalysis extends PredicateHelper {
                   checkColumnNotExists("add", add.fieldNames(), table.schema)
                 }
                 val parent = findParentStruct("add", add.fieldNames())
-                positionArgumentExists(add.position(), parent)
+                add.position match {
+                  case after: After =>
+                    // Handle the case where column position is referencing new columns being added.
+                    if (!colsToAdd.contains(add.fieldNames().init :+ after.column)) {
+                      positionArgumentExists(add.position(), parent)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: shall we just add one more parameter to `positionArgumentExists` for the newly added columns?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala]: just want to make sure all the caller side of `positionArgumentExists` take care of newly added columns.

**Review by cloud-fan — APPROVED** [review]
LGTM except one comment

**Comment by cloud-fan** [comment]
retest this please

**Comment by brkyvz** [comment]
LGTM as well!

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Pull request closed without merging**